# Table of Contents
 <p><div class="lev1 toc-item"><a href="#Introduction" data-toc-modified-id="Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introduction</a></div><div class="lev1 toc-item"><a href="#Data-Acquisition" data-toc-modified-id="Data-Acquisition-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data Acquisition</a></div><div class="lev2 toc-item"><a href="#Acquire-data-from-open-street-maps" data-toc-modified-id="Acquire-data-from-open-street-maps-21"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Acquire data from open street maps</a></div><div class="lev2 toc-item"><a href="#Transform-xml-to-geojson" data-toc-modified-id="Transform-xml-to-geojson-22"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Transform xml to geojson</a></div><div class="lev1 toc-item"><a href="#Data-Cleaning" data-toc-modified-id="Data-Cleaning-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Data Cleaning</a></div><div class="lev2 toc-item"><a href="#Iteration-one" data-toc-modified-id="Iteration-one-31"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Iteration one</a></div><div class="lev3 toc-item"><a href="#Explore-data" data-toc-modified-id="Explore-data-311"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Explore data</a></div><div class="lev3 toc-item"><a href="#Find-Problem-in-Data" data-toc-modified-id="Find-Problem-in-Data-312"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Find Problem in Data</a></div><div class="lev3 toc-item"><a href="#Measure-the-problem" data-toc-modified-id="Measure-the-problem-313"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Measure the problem</a></div><div class="lev3 toc-item"><a href="#Create-solution" data-toc-modified-id="Create-solution-314"><span class="toc-item-num">3.1.4&nbsp;&nbsp;</span>Create solution</a></div><div class="lev3 toc-item"><a href="#Apply-solution" data-toc-modified-id="Apply-solution-315"><span class="toc-item-num">3.1.5&nbsp;&nbsp;</span>Apply solution</a></div><div class="lev3 toc-item"><a href="#Measure-result" data-toc-modified-id="Measure-result-316"><span class="toc-item-num">3.1.6&nbsp;&nbsp;</span>Measure result</a></div><div class="lev2 toc-item"><a href="#Iteration-2" data-toc-modified-id="Iteration-2-32"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Iteration 2</a></div><div class="lev3 toc-item"><a href="#Explore-data" data-toc-modified-id="Explore-data-321"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Explore data</a></div><div class="lev3 toc-item"><a href="#Find-Problem-in-Data" data-toc-modified-id="Find-Problem-in-Data-322"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Find Problem in Data</a></div><div class="lev3 toc-item"><a href="#Measure-the-problem" data-toc-modified-id="Measure-the-problem-323"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Measure the problem</a></div><div class="lev3 toc-item"><a href="#Create-solution" data-toc-modified-id="Create-solution-324"><span class="toc-item-num">3.2.4&nbsp;&nbsp;</span>Create solution</a></div><div class="lev3 toc-item"><a href="#Apply-solution" data-toc-modified-id="Apply-solution-325"><span class="toc-item-num">3.2.5&nbsp;&nbsp;</span>Apply solution</a></div><div class="lev3 toc-item"><a href="#Measure-result" data-toc-modified-id="Measure-result-326"><span class="toc-item-num">3.2.6&nbsp;&nbsp;</span>Measure result</a></div><div class="lev1 toc-item"><a href="#Data-Exploration-with-mongo" data-toc-modified-id="Data-Exploration-with-mongo-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Data Exploration with mongo</a></div><div class="lev2 toc-item"><a href="#Loading-data-into-mongo" data-toc-modified-id="Loading-data-into-mongo-41"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Loading data into mongo</a></div><div class="lev2 toc-item"><a href="#Mongo-queries" data-toc-modified-id="Mongo-queries-42"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Mongo queries</a></div><div class="lev2 toc-item"><a href="#Using-Geopandas" data-toc-modified-id="Using-Geopandas-43"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Using Geopandas</a></div><div class="lev2 toc-item"><a href="#Mapping-Starbucks-with-Folium" data-toc-modified-id="Mapping-Starbucks-with-Folium-44"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Mapping Starbucks with Folium</a></div><div class="lev2 toc-item"><a href="#Mapping-Parks-and-Greenspaces-with-Folium" data-toc-modified-id="Mapping-Parks-and-Greenspaces-with-Folium-45"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Mapping Parks and Greenspaces with Folium</a></div><div class="lev3 toc-item"><a href="#Now-let's-consider-pitches,-parks,-playgrounds" data-toc-modified-id="Now-let's-consider-pitches,-parks,-playgrounds-451"><span class="toc-item-num">4.5.1&nbsp;&nbsp;</span>Now let's consider pitches, parks, playgrounds</a></div><div class="lev3 toc-item"><a href="#Closest-Park" data-toc-modified-id="Closest-Park-452"><span class="toc-item-num">4.5.2&nbsp;&nbsp;</span>Closest Park</a></div>

# Introduction

The primary goal of this project is to use python and mongo to clean and analyze larger-than-RAM. 

This project departs from many suggestions made in the course.

The first departure is _not_ processing an XML file to generate an appropriate JSON file which we can clean and then enter into a mongo database for subsequent analysis.
The reason for this is manifold but foremost is the fact that while the data does still need to be cleaned, the geojson format is already a mongo friendly standard and permits use of a very large set of tools and that there are already utilities that make this conversion. Thus we use a premade tool for making this conversion: `osmtogeojson`. 

Another reason for not processing XML any more than necessary is the library `dask`. This library permits out-of-core and parallel execution, thus reducing processing time by close to the reciprocal of the number of processing cores on has available while avoiding the constraints of RAM for dealing with large data sets. However, while `dask` handle json well, it is not currently adapted for XML.

The format of geojson is desirable because:
- its json, so we can use `dask`
- mongo has spatial indexing which use its objects
- the library `geopandas` reads geojson and provides a multitude of geometric operations all with subclasses derived from pandas' `DataFrame` and `Series`.
- the library `folium` integrates geojson allowing geocoding and choropleths 

Finally, it is important to have a driving reason for why the task is being done, or at least one should be found. In cleaning the data of even a moderately sized city such
as Edmonton the amount of possible work is as large as the granularity of information is small. Therefore, one should try to establish as quickly as possible what exactly one
is looking for. In this case the driving question will be where in Edmonton can I have a Starbuck's espresso beside a park. 



# Data Acquisition

## Acquire data from open street maps

By looking at the file contents, we see that it was produced from the OverPass API using the parameters <bounds minlat="53.4271" minlon="-113.666" maxlat="53.6495" maxlon="-113.34"/>
http://overpass-api.de/api/map?bbox=-113.7538,53.4253,-113.1952,53.6514 . Clicking on this linking downloads a file call `map` which we will place in our notebook directory. 



## Transform xml to geojson

We rename the file `map` to `map.osm` in order for `osmtogeojson` to work properly.

In [11]:
!node --max_old_space_size=8192 `which osmtogeojson` map.osm > map.geojson

Because `osmtogeojson` produces a _pretty-printed_ file, which `dask` cannot read, we transform it into a _JSON lines_ format, which it can.

In [1]:
!head map.geojson

{
"type": "FeatureCollection",
"features": [
{
  "type": "Feature",
  "id": "way/23234822",
  "properties": {
    "timestamp": "2012-05-24T23:12:30Z",
    "version": "6",
    "changeset": "11694406",


We use the library `ijson` to transform a pretty-printed file to a json-lines format (one document per line). However, to do so we need to subclass `json.JSONEncoder` to correctly encode the `Decimal` class that `ijson` creates with floats.

In [3]:
import json
import decimal

class DecimalEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, decimal.Decimal):
#             print("got dec")
            return float(obj)
        # Let the base class default method raise the TypeError
        return json.JSONEncoder.default(self, obj)

In [4]:
import json
import ijson

fi =open('map.geojson')
fo = open('ed_features', mode='w')
features = ijson.items(fi, 'features.item')
# features = (o for o in objects if o['type'] == 'city')

i = 0 
for feature in features:
    try:
        print(json.dumps(feature, cls=DecimalEncoder), file=fo)
    except:
        print(feature)
   

# Data Cleaning


In this section, after we load the appropriate libaries, we will repeat the following sequence: 
- examine data / identify problem
- create measure of problem
- create cleaning method
- apply cleaning method
- measure problem again

Load Librairies

In [43]:
from pprint import pprint
import re
import pandas as pd
# import codecs
import json
# import datetime
from collections import defaultdict
import dask.bag as db
# import sys
import dask
from toolz import take
import toolz
# import pickle
# import pymongo
import graphviz # for dask computation vizualizations
import matplotlib # for basic graphics
import seaborn as sns # for easier graphics

## Iteration one

### Explore data

In [21]:
# we up the blocksize until the number of partitions was at least  8 (my laptops number of cores)
import math
# request to dask developpers - make option to Bag.read_text() to optimize for number of cores cores #!!
i1_b = db.read_text('ed_features', blocksize=math.floor(0.9 * 2**25))#.map(json.loads)

In [22]:
i1_b.npartitions

8

In [25]:
# convert bag of json strings into dictionary objects
j1_b = i1_b.map(json.loads)

We see below that the root keys are 'geometry', 'id', 'properties' and 'type'

In [30]:
set([i for d in j1_b.take(1000) for i in d])

{'geometry', 'id', 'properties', 'type'}

We look at examples of each of these:

In [32]:
j1_b.take(1)[0]['id']

'way/23234822'

In [33]:
j1_b.take(1)[0]['type']

'Feature'

In [34]:
j1_b.take(1)[0]['properties']

{'changeset': '11694406',
 'id': 'way/23234822',
 'natural': 'water',
 'timestamp': '2012-05-24T23:12:30Z',
 'uid': '239998',
 'user': 'Sundance',
 'version': '6'}

In [365]:
print(j1_b.take(5)[-1]['geometry'])

{'type': 'Polygon', 'coordinates': [[[-113.4659956, 53.5683704], [-113.466061, 53.5683707], [-113.4660612, 53.5683557], [-113.4660671, 53.5683557], [-113.4660682, 53.5682687], [-113.4664456, 53.5682704], [-113.4664463, 53.5682164], [-113.4665029, 53.5682167], [-113.466503, 53.5682043], [-113.4667136, 53.5682052], [-113.4667166, 53.5679699], [-113.4670317, 53.5679713], [-113.4670373, 53.5675276], [-113.4658727, 53.5675224], [-113.4658682, 53.5678819], [-113.4658802, 53.567882], [-113.465879, 53.5679748], [-113.4658983, 53.5679748], [-113.4658935, 53.5683571], [-113.4659958, 53.5683575], [-113.4659956, 53.5683704]], [[-113.4665258, 53.5679376], [-113.4664905, 53.5679375], [-113.4664904, 53.5679477], [-113.4662249, 53.567947], [-113.466225, 53.567937], [-113.4661832, 53.5679369], [-113.4661851, 53.5676415], [-113.4665278, 53.5676424], [-113.4665258, 53.5679376]]]}


The key 'properties' looks like it will have most user information in it so we pursue its characteristics. We collect all keys used within the properties sub-document below:

In [93]:
%%time 
c_d = defaultdict(int)
#for d in j1_b.pluck('properties').take(1000) for i in d
for d in j1_b.pluck('properties'):
    for k in d:
        c_d[k] += 1

CPU times: user 3.13 s, sys: 721 ms, total: 3.85 s
Wall time: 18.7 s


In [50]:
pd.DataFrame.from_dict(c_d, orient='index' ).sort_values(0, ascending=False).head(30)

0
user                          467414
version                       467414
timestamp                     467414
id                            467414
uid                           467414
changeset                     467414
building                      281033
source                        142042
addr:street                    81450
addr:housenumber               81432
addr:city                      81027
highway                        49699
addr:interpolation             37942
name                           31551
lanes                          15511
is_in                          10506
attribution                     9930
geobase:acquisitionTechnique    9887
service                         8817
oneway                          7398
surface                         5312
lit                             4544
power                           3923
landuse                         3901
amenity                         3897
maxspeed                        3398
geobase:datasetName             2581
geobase:uuid                    2534
leisure                         2227
foot                            2107

We compare the alternative method below which uses the dask functional approach more consistently. The result is about a 700% speed-up. (18.7s / 2.64s) 

In [243]:
def f1(dd,d):
    for k in d:
        dd[k] +=1 
    return dd

def d2(dda,ddb):
    for k in dda:
        ddb[k] += dda[k]
    return ddb

%time props_df = pd.DataFrame.from_dict(j1_b.pluck('properties').fold(f1,d2,defaultdict(int)).compute(), orient="index").sort_values(0, ascending=False); props_df   

CPU times: user 54.1 ms, sys: 94.1 ms, total: 148 ms
Wall time: 2.92 s


In [244]:
props_df

0
version                       467414
timestamp                     467414
user                          467414
changeset                     467414
uid                           467414
id                            467414
building                      281033
source                        142042
addr:street                    81450
addr:housenumber               81432
addr:city                      81027
highway                        49699
addr:interpolation             37942
name                           31551
lanes                          15511
is_in                          10506
attribution                     9930
geobase:acquisitionTechnique    9887
service                         8817
oneway                          7398
surface                         5312
lit                             4544
power                           3923
landuse                         3901
amenity                         3897
maxspeed                        3398
geobase:datasetName             2581
geobase:uuid                    2534
leisure                         2227
foot                            2107
...                              ...
recycling:concrete                 1
b                                  1
cinema:3D                          1
it                                 1
phone_1                            1
hour_on                            1
max_age                            1
fuel:propane_tank                  1
social_facility_for                1
category                           1
note:ko                            1
alt_name:ko                        1
recycling:paper_packaging          1
roof:height                        1
name:ja                            1
recycling:magazines                1
converstruction                    1
recycling:newspaper                1
operator4                          1
diet:vegetarian                    1
gns:v:xx:full_name_nd              1
recycling:scrap_metal              1
beer_garden                        1
service_times                      1
amenity_3                          1
recycling:wood                     1
building:flats                     1
line                               1
leisure_1                          1
alt_name:zh_pinyin                 1

[532 rows x 1 columns]

From the above table we see the most frequent properities used. In our case we will focus on 'amenity'.

In [96]:
j1_b.pluck('properties').pluck('amenity', '').frequencies().topk(
    10, lambda x: x[1]).to_dataframe().compute()

0       1
0                    463517
1           parking    1127
2         fast_food     420
3        restaurant     419
4            school     328
5              fuel     218
6              cafe     175
7              bank     170
8  place_of_worship     135
9             bench      92

In the above table we see in the first row than there some 463000 documents without an amenity key. Among those that do, there are 175 'cafe'. We will now focus on these.

### Find Problem in Data

From the data below we notice that there are apparently more _Tim Hortons_ than any other kind of café, until we notice that there are two ways of entering _Starbucks_.

Thus we have identified a problem to be resolved. Our measure of this problem will be the same method with which we found it.

Below we create a function to unify the disparate ways of writing Starbucks.

In [97]:
j1_b.pluck('properties').filter(   # get properties sub-doc 
    lambda d: d.get('amenity', '') == 'cafe').pluck(  # filter to get those in which amenity == cafe
        'name', '__').frequencies().topk(              # from those documents take the names, and get the frequencies of those names
            10, lambda x: x[1]).to_dataframe().compute() # take the 10 most frequent and make a dataframe from them

0   1
0       Tim Hortons  35
1        Second Cup  25
2         Starbucks  19
3  Starbucks Coffee  17
4     Booster Juice   4
5     Orange Julius   4
6                __   3
7    The Second Cup   2
8        Good Earth   2
9   Dream Tea House   2

### Measure the problem

We will abstract our count of cafés so that we can use it to measure our success on our new data set.

In [113]:
def cafe_count( j_b ):
    df = j_b.pluck('properties').filter(   # get properties sub-doc 
    lambda d: d.get('amenity', '') == 'cafe').pluck(  # filter to get those in which amenity == cafe
        'name', '__').frequencies().topk(              # from those documents take the names, and get the frequencies of those names
            10, lambda x: x[1]).to_dataframe().compute() # take the 10 most frequent and make a dataframe from them
    return df

We verify that it works as expected:

In [114]:
cafe_count(j1_b)

0   1
0       Tim Hortons  35
1        Second Cup  25
2         Starbucks  19
3  Starbucks Coffee  17
4     Booster Juice   4
5     Orange Julius   4
6                __   3
7    The Second Cup   2
8        Good Earth   2
9   Dream Tea House   2

### Create solution

In [ ]:
re_sb = re.compile('starbucks', re.IGNORECASE)


def save_starbucks(d):
    dp = d['properties']
    s = dp.get('name', '')
    if re_sb.search(s):
        dp['name'] = 'Starbucks'
    return d

### Apply solution

We map this function on to the documents in our `j1_b` dask bag. It should be observed that are _latent execution graphs_, and not actually calculated until they recieve a message such as `compute`, `take`, or `to_text`.

In [107]:
j11_b = j1_b.map(save_starbucks)

We then use the same method to observe whether our new bag is correctly amended and it appears that it has as our beloved purveyor of espresso is properly credited with all its stores. (see below)

### Measure result

In [115]:
cafe_count(j11_b)

0   1
0         Starbucks  36
1       Tim Hortons  35
2        Second Cup  25
3     Booster Juice   4
4     Orange Julius   4
5                __   3
6    The Second Cup   2
7        Good Earth   2
8   Dream Tea House   2
9  Transcend Coffee   2

By the fact that there is only one listing for Starbucks, we are satisfied that this anomaly has been cleaned.

We will write out our amended json and then read it in to mongo using `mongoimport`.

The following command will write out 8 json files, one for each partition of the dask bag.

In [379]:
j11_b.map(json.dumps).to_textfiles('./data_*.json')

In [380]:
!ls -lh data_?.json

-rw-r--r--  1 mannalytics  staff    29M  2 Feb 08:47 data_0.json
-rw-r--r--  1 mannalytics  staff    29M  2 Feb 08:47 data_1.json
-rw-r--r--  1 mannalytics  staff    29M  2 Feb 08:47 data_2.json
-rw-r--r--  1 mannalytics  staff    29M  2 Feb 08:47 data_3.json
-rw-r--r--  1 mannalytics  staff    29M  2 Feb 08:47 data_4.json
-rw-r--r--  1 mannalytics  staff    29M  2 Feb 08:47 data_5.json
-rw-r--r--  1 mannalytics  staff    29M  2 Feb 08:47 data_6.json
-rw-r--r--  1 mannalytics  staff    12M  2 Feb 08:47 data_7.json


## Iteration 2

### Explore data

In [381]:
# we read in the files that we wrote in the previous iteration
i2_b = db.read_text('data_*.json')

In [382]:
i2_b.npartitions

8

In [383]:
# convert bag of json strings into dictionary objects
j2_b = i2_b.map(json.loads)

In [385]:
j2_b.pluck('properties').pluck('amenity', '').frequencies().topk(
    10, lambda x: x[1]).to_dataframe().compute()

0       1
0                    463517
1           parking    1127
2         fast_food     420
3        restaurant     419
4            school     328
5              fuel     218
6              cafe     175
7              bank     170
8  place_of_worship     135
9             bench      92

### Find Problem in Data

After perusing a few of the amenities we find that the amenity 'fuel' has some problems. There are synonymous representations:
- '7-Eleven / 7-11', 
- 'FasGas / Fas Gas / Fast Gas Plus', 
- 'Petro Canada / PetroCanada / Petro-Canada'  
- 'Safeway Gas Bar / Safeway Gasoline / Safeway Gas', 
- 'Hughes / hughes'
- 'Esso / Esso Gas Bar'

In [393]:
j2_b.pluck('properties').filter(lambda d: d.get('amenity','') == 'fuel').pluck('name','').frequencies().topk(
    30, lambda x: x[1]).to_dataframe().compute()

0   1
0                       Esso  30
1                      Husky  28
2                      Shell  25
3               Petro Canada  19
4                             17
5                     Hughes  12
6                   7-Eleven  10
7                       Domo   9
8               Petro-Canada   8
9                       7-11   6
10        Superstore Gas Bar   6
11           Safeway Gas Bar   5
12           Costco Gasoline   4
13             Canadian Tire   3
14                  Flying J   2
15               FasGas Plus   2
16              Fas Gas Plus   2
17               PetroCanada   2
18                   Safeway   2
19                    Centex   2
20       Convenience For You   1
21          Safeway Gasoline   1
22                    hughes   1
23             Fast Gas Plus   1
24  Esso Gas Bar/Tim Hortons   1
25                   Gas Bar   1
26                   Chevron   1
27             Shell Express   1
28               Safeway Gas   1
29             Triple S Auto   1

### Measure the problem

We will abstract our count of cafés so that we can use it to measure our success on our new data set.

In [396]:
def gas_count( j_b, top_n = 10 ):
    df = j_b.pluck('properties').filter(   # get properties sub-doc 
    lambda d: d.get('amenity', '') == 'fuel').pluck(  # filter to get those in which amenity == cafe
        'name', '__').frequencies().topk(              # from those documents take the names, and get the frequencies of those names
            top_n, lambda x: x[1]).to_dataframe().compute() # take the 10 most frequent and make a dataframe from them
    return df

We verify that it works as expected:

In [417]:
before = gas_count(j1_b, top_n = 30); before

0   1
0                       Esso  30
1                      Husky  28
2                      Shell  25
3               Petro Canada  19
4                         __  17
5                     Hughes  12
6                   7-Eleven  10
7                       Domo   9
8               Petro-Canada   8
9                       7-11   6
10        Superstore Gas Bar   6
11           Safeway Gas Bar   5
12           Costco Gasoline   4
13             Canadian Tire   3
14                  Flying J   2
15               FasGas Plus   2
16              Fas Gas Plus   2
17               PetroCanada   2
18                   Safeway   2
19                    Centex   2
20       Convenience For You   1
21          Safeway Gasoline   1
22                    hughes   1
23             Fast Gas Plus   1
24  Esso Gas Bar/Tim Hortons   1
25                   Gas Bar   1
26                   Chevron   1
27             Shell Express   1
28               Safeway Gas   1
29             Triple S Auto   1

### Create solution

In [414]:


re_dict = { re.compile('7', re.IGNORECASE) : '7-Eleven',
           re.compile('Fas.*Gas', re.IGNORECASE) : 'FasGas',
           re.compile('Petro.*Canada', re.IGNORECASE) : 'Petro Canada',
           re.compile('Safeway', re.IGNORECASE) : 'Safeway',
           re.compile('Hughes', re.IGNORECASE) : 'Hughes',
           re.compile('Esso', re.IGNORECASE) : 'Hughes'}


def unify_gas(d):
    dp = d['properties']
    s = dp.get('name', '')
    for k_re, name in re_dict.items():
        if k_re.search(s):
            dp['name'] = name
    return d

### Apply solution

In [415]:
j21_b = j2_b.map(unify_gas)

We then use the same method to observe whether our new bag is correctly amended and it appears that it has as our beloved purveyor of espresso is properly credited with all its stores. (see below)

### Measure result

In the result below we see most, but not all, of our problems resolved

In [448]:
after = gas_count(j21_b,top_n = 30)

In [459]:
_df = pd.concat([before, after],axis=1)
_df.columns = ["Before",'count_b',"After","count_a"]
_df.count_a = _df.count_a.fillna(0).astype(int)
_df

Before  count_b                                 After  \
0                       Esso       30                                Hughes   
1                      Husky       28                          Petro Canada   
2                      Shell       25                                 Husky   
3               Petro Canada       19                                 Shell   
4                         __       17                                    __   
5                     Hughes       12                              7-Eleven   
6                   7-Eleven       10                               Safeway   
7                       Domo        9                                  Domo   
8               Petro-Canada        8                    Superstore Gas Bar   
9                       7-11        6                                FasGas   
10        Superstore Gas Bar        6                       Costco Gasoline   
11           Safeway Gas Bar        5                         Canadian Tire   
12           Costco Gasoline        4                              Flying J   
13             Canadian Tire        3                                Centex   
14                  Flying J        2                   Convenience For You   
15               FasGas Plus        2                                 Co-Op   
16              Fas Gas Plus        2                 No Frills Gas Station   
17               PetroCanada        2                              Gas Stop   
18                   Safeway        2                               Gas Bar   
19                    Centex        2    City of Edmonton Central Fuel Site   
20       Convenience For You        1                               Chevron   
21          Safeway Gasoline        1                         Shell Express   
22                    hughes        1                 Canadian Tire Gas Bar   
23             Fast Gas Plus        1                         Triple S Auto   
24  Esso Gas Bar/Tim Hortons        1  Real Canadian Superstore Gas Station   
25                   Gas Bar        1               Mac's Convenience Store   
26                   Chevron        1                                  Coop   
27             Shell Express        1                                 Mac's   
28               Safeway Gas        1                            Petro-Pass   
29             Triple S Auto        1                                   NaN   

    count_a  
0        46  
1        29  
2        28  
3        25  
4        17  
5        17  
6        10  
7         9  
8         6  
9         5  
10        4  
11        3  
12        2  
13        2  
14        1  
15        1  
16        1  
17        1  
18        1  
19        1  
20        1  
21        1  
22        1  
23        1  
24        1  
25        1  
26        1  
27        1  
28        1  
29        0

On visual inspection this looks corrected - it is easy to imagine that a computational approach might be necessary if we had more items to compare.

# Data Exploration with mongo

## Loading data into mongo

We create a pymongo client object to communicate with the mongo database.

In [118]:
from pymongo import MongoClient
client = MongoClient('localhost:27017')
mdb = client['examples']

We want a clean collection so we drop any pre-existing data.

In [119]:
mdb.p33.drop()

Finally, we use `mongoimport` to read in the 8 json files we create above.

In [120]:
%%bash
ls ?.json | while read col; do
    mongoimport --db examples --collection p33 -j 8 --type json --file $col;
done

2017-01-31T20:54:01.654-0700	connected to: localhost
2017-01-31T20:54:03.936-0700	imported 49202 documents
2017-01-31T20:54:03.960-0700	connected to: localhost
2017-01-31T20:54:05.857-0700	imported 56009 documents
2017-01-31T20:54:05.880-0700	connected to: localhost
2017-01-31T20:54:07.877-0700	imported 61973 documents
2017-01-31T20:54:07.898-0700	connected to: localhost
2017-01-31T20:54:09.906-0700	imported 62836 documents
2017-01-31T20:54:09.929-0700	connected to: localhost
2017-01-31T20:54:11.906-0700	imported 61785 documents
2017-01-31T20:54:11.930-0700	connected to: localhost
2017-01-31T20:54:14.056-0700	imported 66069 documents
2017-01-31T20:54:14.079-0700	connected to: localhost
2017-01-31T20:54:16.141-0700	imported 74989 documents
2017-01-31T20:54:16.164-0700	connected to: localhost
2017-01-31T20:54:17.083-0700	imported 34551 documents


## Mongo queries

We are now at a point where we can make queries on our database.

In [126]:
for d in mdb.p33.find({"properties.landuse": {"$exists": 1}},{"properties.landuse":1}):
    

{'_id': ObjectId('58915bdaaaf4167ceca9fb75'),
 'properties': {'landuse': 'grass'}}

In [130]:
list(mdb.p33.aggregate([{"$sortByCount": "$properties.landuse"}, {"$limit": 30}]))
    

[{'_id': None, 'count': 463513},
 {'_id': 'residential', 'count': 2351},
 {'_id': 'retail', 'count': 449},
 {'_id': 'grass', 'count': 331},
 {'_id': 'commercial', 'count': 248},
 {'_id': 'industrial', 'count': 225},
 {'_id': 'basin', 'count': 56},
 {'_id': 'forest', 'count': 53},
 {'_id': 'recreation_ground', 'count': 47},
 {'_id': 'quarry', 'count': 30},
 {'_id': 'reservoir', 'count': 24},
 {'_id': 'cemetery', 'count': 19},
 {'_id': 'construction', 'count': 19},
 {'_id': 'farmland', 'count': 10},
 {'_id': 'military', 'count': 6},
 {'_id': 'meadow', 'count': 4},
 {'_id': 'allotments', 'count': 4},
 {'_id': 'protected_area', 'count': 3},
 {'_id': 'landfill', 'count': 3},
 {'_id': 'civic_admin', 'count': 2},
 {'_id': 'traffic_island', 'count': 2},
 {'_id': 'public transportation', 'count': 1},
 {'_id': 'greenhouse_horticulture', 'count': 1},
 {'_id': 'mixed', 'count': 1},
 {'_id': 'brownfield', 'count': 1},
 {'_id': 'greenfield', 'count': 1},
 {'_id': 'conservation', 'count': 1},
 {'_id'

Of interest above is 'grass','forest' and possibly 'recreation_ground'. 

In [139]:
list(mdb.p33.aggregate([{"$match": {"properties.landuse" : {"$in": ["grass","forest","recreation_ground"]}}}, {"$sortByCount": "$properties.landuse"}, {"$limit": 30}]))
    

[{'_id': 'grass', 'count': 331},
 {'_id': 'forest', 'count': 53},
 {'_id': 'recreation_ground', 'count': 47}]

In [185]:
mdb.p33.count( {"properties.name": "Starbucks" })

40

In [143]:
mdb.p33.count({"properties.landuse" : {"$in": ["grass","forest","recreation_ground"]}})

431

In [144]:
mdb.p33.count({"$or": [{"properties.landuse" : {"$in": ["grass","forest","recreation_ground"]}},  {"properties.name": "Starbucks" }]})

471

## Using Geopandas

In [147]:
import geopandas as gpd

In [149]:
### mongodb and geopandas play together!!
iter = mdb.p33.find({"$or": [{"properties.landuse" : {"$in": ["grass","forest","recreation_ground"]}},  {"properties.name": "Starbucks" }]})
gdf = gpd.GeoDataFrame.from_features(iter)

In [150]:
gdf.head()

addr:city addr:housename addr:housenumber addr:postcode addr:province  \
0       NaN            NaN              NaN           NaN           NaN   
1       NaN            NaN              NaN           NaN           NaN   
2       NaN            NaN              NaN           NaN           NaN   
3       NaN            NaN              NaN           NaN           NaN   
4       NaN            NaN              NaN           NaN           NaN   

  addr:street amenity area bicycle building ...  smoking source sport surface  \
0         NaN     NaN  NaN     NaN      NaN ...      NaN    NaN   NaN     NaN   
1         NaN     NaN  NaN     NaN      NaN ...      NaN    NaN   NaN     NaN   
2         NaN     NaN  NaN     NaN      NaN ...      NaN    NaN   NaN     NaN   
3         NaN     NaN  NaN     NaN      NaN ...      NaN    NaN   NaN     NaN   
4         NaN     NaN  NaN     NaN      NaN ...      NaN    NaN   NaN     NaN   

              timestamp          type      uid            user version wifi  
0  2016-09-03T17:25:24Z  multipolygon  3274426          yegbin       2  NaN  
1  2016-10-16T03:33:40Z  multipolygon  3274426          yegbin       1  NaN  
2  2016-02-12T06:17:13Z  multipolygon  1694420  Little Brother       1  NaN  
3  2016-09-23T15:28:28Z  multipolygon  3274426          yegbin       1  NaN  
4  2016-10-16T03:33:40Z  multipolygon  3274426          yegbin       1  NaN  

[5 rows x 45 columns]

In [151]:
gdf.columns.T

Index(['addr:city', 'addr:housename', 'addr:housenumber', 'addr:postcode',
       'addr:province', 'addr:street', 'amenity', 'area', 'bicycle',
       'building', 'changeset', 'created_by', 'cuisine', 'foot', 'geometry',
       'golf', 'highway', 'horse', 'id', 'internet_access', 'landcover',
       'landuse', 'layer', 'leaf_cycle', 'leaf_type', 'leisure', 'level',
       'name', 'note', 'oneway', 'opening_hours', 'phone', 'place', 'service',
       'shop', 'smoking', 'source', 'sport', 'surface', 'timestamp', 'type',
       'uid', 'user', 'version', 'wifi'],
      dtype='object')

The CRS (Coordinate Reference System) needs to be set in order to calculate areas in units such a meters.

[epsg:3395 description](http://epsg-registry.org/report.htm?type=selection&entity=urn:ogc:def:crs:EPSG::3395&reportDetail=short&style=urn:uuid:report-style:default-with-code&style_name=OGP%20Default%20With%20Code&title=EPSG:3395) uses meters


[epsg:4326 description](https://nsidc.org/data/atlas/epsg_4326.html) uses degrees

In [187]:
gdf.crs = {'init': 'epsg:4326', 'no_defs': True}
gdf.loc[gdf.landuse.isin(['grass','forest']), ['landuse','geometry']].to_crs({'init': 'epsg:3395'}).area

0      955101.711344
1       11152.403134
2       33734.277725
3       30104.316550
4        1510.748314
6       19179.814746
7       10099.030055
8        6810.182690
9      829536.269136
10      37188.019559
11      13663.380438
13       9890.555829
16        278.149317
21        927.817790
24      29953.950749
28      24686.753919
30      26198.613249
36      70491.940885
37      10978.133358
40      59956.150168
41     174601.953976
62      52784.318929
64      39405.048195
65       1267.241929
66       4553.359278
67      20050.839657
68       2478.230599
69      23699.910653
70      41601.704173
71       8910.540123
           ...      
407     14923.690965
408      1019.015080
409     18674.578157
410      1500.427424
411     33310.265895
412      1474.224958
413      2375.975651
414      2282.682911
415      2762.255024
416      2172.389769
417      1143.367051
418     11191.572440
419       664.144859
420     66140.472421
421    117124.234219
422     19863.866599
423      7813

In [197]:
gdf.crs = {'init': 'epsg:4326', 'no_defs': True}
starbucks_locations = gdf.loc[gdf.name == "Starbucks",:].centroid

## Mapping Starbucks with Folium

In [236]:
import folium
from folium import plugins

e_map = folium.Map(location=[53.5227229, -113.5116757], zoom_start=12)
marker_cluster = folium.MarkerCluster().add_to(e_map)

for d in starbucks_locations:  
    folium.Marker(
        [d.y, d.x], icon=folium.Icon(color='green'),
        popup='Starbucks').add_to(marker_cluster)

# e_map.save('e_map.html')

In [237]:
e_map

In [181]:
gdf.crs

{'init': 'epsg:4326', 'no_defs': True}

## Mapping Parks and Greenspaces with Folium

In [217]:
# cut and pasted from http://andrewgaidus.com/leaflet_webmaps_python/ - thank-you Andrew Gaidus!

import pysal as ps
def add_choropleth(mapobj, gdf, id_field, value_field, fill_color = 'YlOrRd', fill_opacity = 0.6, 
                    line_opacity = 0.2, num_classes = 5, classifier = 'Fisher_Jenks'):
    #Allow for 3 Pysal map classifiers to display data
    #Generate list of breakpoints using specified classification scheme. List of breakpoint will be input to choropleth function
    if classifier == 'Fisher_Jenks':
        threshold_scale=ps.esda.mapclassify.Fisher_Jenks(gdf[value_field], k = num_classes).bins.tolist()
    if classifier == 'Equal_Interval':
        threshold_scale=ps.esda.mapclassify.Equal_Interval(gdf[value_field], k = num_classes).bins.tolist()
    if classifier == 'Quantiles':
        threshold_scale=ps.esda.mapclassify.Quantiles(gdf[value_field], k = num_classes).bins.tolist()
    
    #Convert the GeoDataFrame to WGS84 coordinate reference system
    gdf_wgs84 = gdf.to_crs({'init': 'epsg:4326'})
    
    #Call Folium choropleth function, specifying the geometry as a the WGS84 dataframe converted to GeoJSON, the data as 
    #the GeoDataFrame, the columns as the user-specified id field and and value field.
    #key_on field refers to the id field within the GeoJSON string
    mapobj.choropleth(geo_str = gdf_wgs84.to_json(), data = gdf,
                columns = [id_field, value_field], key_on = 'feature.properties.{}'.format(id_field),
                fill_color = fill_color, fill_opacity = fill_opacity, line_opacity = line_opacity,  
                threshold_scale = threshold_scale)
    return mapobj

In [249]:
list(mdb.p33.aggregate([{"$sortByCount": "$properties.leisure"}, {"$limit": 10}]))
    

[{'_id': None, 'count': 465187},
 {'_id': 'pitch', 'count': 879},
 {'_id': 'park', 'count': 713},
 {'_id': 'playground', 'count': 330},
 {'_id': 'ice_rink', 'count': 102},
 {'_id': 'sports_centre', 'count': 44},
 {'_id': 'picnic_table', 'count': 43},
 {'_id': 'track', 'count': 22},
 {'_id': 'golf_course', 'count': 21},
 {'_id': 'swimming_pool', 'count': 20}]

In [149]:
### mongodb and geopandas play together!!
iter = mdb.p33.find({"$or": [{"properties.leisure" : {"$in": ["pitch","forest","recreation_ground"]}},  {"properties.name": "Starbucks" }]})
gdf = gpd.GeoDataFrame.from_features(iter)

### Now let's consider pitches, parks, playgrounds

In [254]:
iter = mdb.p33.find({"properties.leisure" : {"$in": ["pitch","park","playground"]}})
ppp_gdf = gpd.GeoDataFrame.from_features(iter)

In [256]:
e_map = folium.Map(location=[53.5227229, -113.5116757], zoom_start=12)
ppp_gdf.crs = {'init': 'epsg:4326', 'no_defs': True}

ppp_gdf.area = ppp_gdf.area.astype(float)
add_choropleth(e_map, ppp_gdf[ppp_gdf.area >0], "id","area", line_opacity=0.8, classifier="Quantiles")

### Closest Park

In this final section we use GeoPandas and Folium to label a map such that Starbucks cafés are marked and the park closest to it is outlined in blue. Additionally, when the markers are clicked it will give the distance to the closest park.

In [331]:
def closest_park(pt, _gdf):
    """ assume that pt is in epsg:3395 CRS """
    # crs epsg:3395 lets us measure distance in metres
    _s = _gdf.to_crs({'init': 'epsg:3395', 'no_defs': True}).distance(pt)
    idx = _s.argmin()
    geo =_gdf.loc[idx,'geometry']
    dist = _s[idx]
    return pd.Series({'point': pt, 'idx':idx,'dist':dist,'geo':geo})

In [310]:
# create a starbucks geoSeries with metric units
sb_metric = starbucks_locations.to_crs({'init': 'epsg:3395', 'no_defs': True})

In [332]:
starParks_s = sb_metric.apply(closest_park, _gdf = ppp_gdf)

In [336]:
starParks_gdf = gpd.GeoDataFrame(starParks_s, geometry="geo", crs={'init': 'epsg:4326', 'no_defs': True})

In [344]:
starParks_gdf['Starbucks'] = starbucks_locations

In [443]:
e_map = folium.Map(location=[53.5227229, -113.5116757], zoom_start=12)
# ppp_gdf.crs = {'init': 'epsg:4326', 'no_defs': True}

# e_map = folium.Map(location=[53.5227229, -113.5116757], zoom_start=12)
marker_cluster = folium.MarkerCluster().add_to(e_map)

def pop_up_str(d):
    distt = starParks_gdf.Starbucks

for i in starParks_gdf.index:  
    d = starParks_gdf.loc[i,'Starbucks']
    popup_str = starParks_gdf.loc[i,'dist']
    folium.Marker(
        [d.y, d.x], icon=folium.Icon(color='#FF1234'),
        popup='Distance to park {:3.0f} m'.format(popup_str)).add_to(marker_cluster)

for i,row in starParks_gdf.iterrows():
    shp_pts = list(row.geo.exterior.coords)
    fol_pts = [(p[1],p[0]) for p in shp_pts]
    poly = folium.PolyLine(fol_pts, color='blue')
#     print(poly)
    poly.add_to(e_map)
#     e_map.add_child(poly)

e_map
# add_choropleth(e_map, starParks_gdf.drop('Starbucks',1), "idx","dist",fill_color = 'RdBu', line_opacity=0.9, classifier="Quantiles")